src/backend/distributed/commands/role.c

```diff
@@ -309,144 +417,118 @@ MakeVariableSetStmt(const char *config)
+	options = lappend(options, makeDefElemInt("bypassrls", role->rolbypassrls));
+	options = lappend(options, makeDefElemInt("connectionlimit", role->rolconnlimit));
+
+	/* load password from heap tuple, use NULL if not set */
+	bool isNull = true;
+	Datum rolPasswordDatum = SysCacheGetAttr(AUTHNAME, tuple, Anum_pg_authid_rolpassword,
+											 &isNull);
 	if (!isNull)
 	{
-		rolPassword = pstrdup(TextDatumGetCString(rolPasswordDatum));
-		stmt->options = lappend(stmt->options, makeDefElem("password",
-														   (Node *) makeString(
-															   rolPassword),
-														   -1));
+		char *rolPassword = pstrdup(TextDatumGetCString(rolPasswordDatum));
+		Node *passwordStringNode = (Node *) makeString(rolPassword);
+		DefElem *passwordOption = makeDefElem("password", passwordStringNode, -1);
+		options = lappend(options, passwordOption);
 	}
 	else
 	{
-		stmt->options = lappend(stmt->options, makeDefElem("password", NULL, -1));
+		options = lappend(options, makeDefElem("password", NULL, -1));
 	}
 
-	Datum rolValidUntilDatum = heap_getattr(tuple, Anum_pg_authid_rolvaliduntil,
-											pgAuthIdDescription, &isNull);
+	/* load valid unitl data from the heap tuple, use default of infinity if not set */
+	Datum rolValidUntilDatum = SysCacheGetAttr(AUTHNAME, tuple,
+											   Anum_pg_authid_rolvaliduntil, &isNull);
+	char *rolValidUntil = "infinity";
 	if (!isNull)
 	{
 		rolValidUntil = pstrdup((char *) timestamptz_to_str(rolValidUntilDatum));
 	}
 
-	stmt->options = lappend(stmt->options, makeDefElem("validUntil",
-													   (Node *) makeString(rolValidUntil),
-													   -1));
+	Node *validUntilStringNode = (Node *) makeString(rolValidUntil);
+	DefElem *validUntilOption = makeDefElem("validUntil", validUntilStringNode, -1);
+	options = lappend(options, validUntilOption);
 
-	return CreateAlterRoleIfExistsCommand(stmt);
+	return options;
 }
 
 
 /*
- * GenerateAlterRoleIfExistsCommandAllRoles creates ALTER ROLE commands
- * that copies all roles from the pg_authid table.
+ * GenerateCreateOrAlterRoleCommand generates ALTER ROLE command that copies a role from
+ * the pg_authid table.
  */
 List *
-GenerateAlterRoleIfExistsCommandAllRoles()
+GenerateCreateOrAlterRoleCommand(Oid roleOid)
 {
-	Relation pgAuthId = heap_open(AuthIdRelationId, AccessShareLock);
-	TupleDesc pgAuthIdDescription = RelationGetDescr(pgAuthId);
-	HeapTuple tuple = NULL;
-	List *commands = NIL;
-	const char *alterRoleQuery = NULL;
+	HeapTuple roleTuple = SearchSysCache1(AUTHOID, ObjectIdGetDatum(roleOid));
+	Form_pg_authid role = ((Form_pg_authid) GETSTRUCT(roleTuple));
 
-#if PG_VERSION_NUM >= PG_VERSION_12
-	TableScanDesc scan = table_beginscan_catalog(pgAuthId, 0, NULL);
-#else
-	HeapScanDesc scan = heap_beginscan_catalog(pgAuthId, 0, NULL);
-#endif
+	CreateRoleStmt *createRoleStmt = NULL;
+	AlterRoleStmt *alterRoleStmt = NULL;
+	if (EnableAlterRolePropagation)
```

Comment:
Should this PR introduce EnableCreateRolePropagation as well in case the CREATE ROLE is somehow problematic / undesirable because some other system wants to create the role at a later time?

Reply:
create role propagation is an enterprise feature, not implemented here.
enterprise ads a block in this function that set the `createRoleStmt`.